### DISCUSSION TITLE
Difference between attention contrib ops

### DISCUSSION THREAD

**Discussion by asyncth** [microsoft/onnxruntime#15325] [Other Q&A] [answered]
What is the difference between `com.microsoft.Attention` and `com.microsoft.MultiHeadAttention` contrib operators? `ContribOperators.md` says that both of them are multi-head attention.

**Comment by hariharans29** [comment, marked as answer]
CC: @tianleiwu

`Attention` only support SelfAttention whereas `MultiHeadAttention` supports self and cross attention

**Reply by asyncth** [reply]
Am I understanding it correctly that `MultiHeadAttention` op doesn't do QKV projection and it expects already projected query, key and value as inputs and that both `Attention` and `MultiHeadAttention` ops don't do output projection?

**Reply by tianleiwu** [reply]
Right. There are also some minor difference in past state input. In Attention, each past input is past_key + past_value, while in MultiHeadAttention, past_key and past_value are separated inputs.

**Reply by asyncth** [reply]
Thanks for help!

**Reply by asyncth** [reply]
Should one always prefer `Attention` over `MultiHeadAttention` for self-attention?

**Reply by tianleiwu** [reply]
Sometime MultiHeadAttention is a better choice like the following:
(1) past_key and past_value are separated in ONNX model.
(2) there is no bias in QKV projection and qkv can be packed.

Right now, you can choose Attention if it is applicable.

**Reply by asyncth** [reply]
> (2) there is no bias in QKV projection and qkv can be packed.

What do you mean? Doesn't `Attention` op handle that case just as well?

Also I'm having trouble understanding `bias` input of `MultiHeadAttention`, is just a value that is chunked into 3 tensors of shapes `(hidden_size)`, `(hidden_size)` and `(v_hidden_size)` and then those tensors are added to query, key and value?

**Comment by tianleiwu** [comment]
>Doesn't Attention op handle that case just as well?

For this case (there is no bias in QKV projection and qkv can be packed), Attention need transpose BS3NH to BSN3H for fp16 fused attention in CUDA, while packed QKV is BSN3H format so MultiHeadAttention might be faster since it saves a transpose.

> Also I'm having trouble understanding bias input of MultiHeadAttention, is just a value that is chunked into 3 tensors of shapes (hidden_size), (hidden_size) and (v_hidden_size) and then those tensors are added to query, key and value?

Yes.

**Reply by asyncth** [reply]
Got it, thanks.
